gtda/curves/features.py

```diff
@@ -0,0 +1,217 @@
+        # instances of FunctionType and the "in" key checks fail
+        except ValueError as ve:
+            end_string = f"which is not in " \
+                         f"{tuple(_AVAILABLE_FUNCTIONS.keys())}."
+            function = params["function"]
+            if ve.args[0].endswith(end_string) \
+                    and isinstance(function, (list, tuple)):
+                params["function"] = [f for f in function
+                                      if isinstance(f, str)]
+                validate_params(params, _hyperparameters, exclude=["n_jobs"])
+            else:
+                raise ve
+
+        if isinstance(self.function, (list, tuple)) \
+                and isinstance(self.function_params, dict):
+            raise TypeError("If `function` is a list/tuple then "
+                            "`function_params` must be a list/tuple of dict, "
+                            "or None.")
+        elif isinstance(self.function, (str, FunctionType)) \
+                and isinstance(self.function_params, (list, tuple)):
+            raise TypeError("If `function` is a string or a callable "
+                            "function then `function_params` must be a dict "
+                            "or None.")
+
+    def fit(self, X, y=None):
+        """Compute :attr:`function_` and :attr:`effective_function_params_`.
+        Then, return the estimator.
+
+        This function is here to implement the usual scikit-learn API and hence
+        work in pipelines.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_channels, n_bins)
+            Input data. Collection of multi-channel curves.
+
+        y : None
+            There is no need for a target in a transformer, yet the pipeline
+            API requires this parameter.
+
+        Returns
+        -------
+        self : object
+
+        """
+        check_array(X, ensure_2d=False, allow_nd=True)
+        if X.ndim != 3:
+            raise ValueError("Input must be 3-dimensional.")
+        self._validate_params()
+
+        self.n_channels_ = X.shape[-2]
+
+        if isinstance(self.function, str):
+            self._function = _implemented_function_recipes[self.function]
+
+            if self.function_params is None:
+                self.effective_function_params_ = {}
+            else:
+                validate_params(self.function_params,
+                                _AVAILABLE_FUNCTIONS[self.function])
+                self.effective_function_params_ = self.function_params.copy()
+
+        else:
+            if isinstance(self.function, FunctionType):
+                self._function = tuple([self.function] * self.n_channels_)
+                if self.function_params is None:
+                    self.effective_function_params_ = \
+                        tuple([{}] * self.n_channels_)
+                else:
+                    self.effective_function_params_ = \
+                        tuple(self.function_params.copy()
```

Comment:
Aha! I thought this would be a bad trap because each entry in `[self.function_params.copy()] * self.n_channels_` refers to the same python object.
Try this:
```
d = {"spam": "eggs"}
t = tuple([d] * 2)
t[0].pop("spam")
```

Reply:
Yes, but we are not modifying them further. As long as the initial `function_params` can't be modified, I it's fine, no?

Reply:
You are right about this, I was being over-cautious here!

Reply:
Done as you suggested